### PR TITLE
Improve support for ranges with nonstandard Integers

### PR DESCRIPTION
**Pull request by tkoolen**
See added `testset`. Makes it so that `UnitRange`, `OneTo`, and `StepRange` properly handle `Integer` subtypes (e.g. `Position <: Integer`) for which _differences_ are of a different type (say `Displacement <: Integer`). Note the `test_broken` for `a : b : c`. I can fix that too if there's support for this PR.

This kind of thing is pretty handy for e.g. `MathOptInterface`, where there's a `VariableIndex` `struct` that's basically a wrapper around an `Int`, and `VariableIndex` ranges would be very handy, but it doesn't really make sense to return a `VariableIndex` from `Base.:-(x::VariableIndex, y::VariableIndex)`.

### CODE REVIEW

File: base/range.jl [hunk 1]
```diff
@@ -374,7 +374,7 @@ julia> step(range(2.5, stop=10.9, length=85))
 ```
 """
 step(r::StepRange) = r.step
-step(r::AbstractUnitRange{T}) where{T} = oneunit(T)
+step(r::AbstractUnitRange{T}) where{T} = one(T) - zero(T)
```
Comment (fredrikekre) [inline base/range.jl]: Should be `oneunit(T) - zero(T)`, no?

Comment (tkoolen) [reply base/range.jl]: Ah yes, I mixed them up and thought that `oneunit` was the multiplicative identity.

**Comment by tkoolen** [comment]
Added support for `start : step : stop` by simply removing the `:` method that promotes `step` (easier than I expected). The whole test suite passes locally.

Also fixes https://github.com/JuliaLang/julia/issues/26619.

**Comment by tkoolen** [comment]
Hmm, there's still an issue with `length` for the case in #26619:

```julia
julia> r = UInt(100) : -1 : 1
0x0000000000000064:-1:0x0000000000000001

julia> length(r)
ERROR: OverflowError: 1 - 100 overflowed for type UInt64
Stacktrace:
 [1] throw_overflowerr_binaryop(::Symbol, ::UInt64, ::UInt64) at ./checked.jl:158
 [2] checked_sub at ./checked.jl:227 [inlined]
 [3] length(::StepRange{UInt64,Int64}) at ./range.jl:404
 [4] top-level scope
```

However,

```julia
julia> r = UInt(100) : -2 : 1
0x0000000000000064:-2:0x0000000000000000

julia> length(r)
0x0000000000000033
```

works fine.

Does anybody know why `length` is implemented the way it is:
https://github.com/JuliaLang/julia/blob/993c551b2fc6ee514df54e3a7959120513f8b62e/base/range.jl#L398-L407

Why are `1` and `-1` special?

**Comment by mbauman** [comment]
My guess is that `div` is expensive, but it's free if LLVM knows the divisor is one.  I think the fix here is to change that last line to divide by `one(S)` instead.

**Comment by tkoolen** [comment]
Actually, I'm not 100% sure that my change to `steprange_last` is correct.

File: base/range.jl [hunk 2]
```diff
@@ -144,7 +143,7 @@ function steprange_last(start::T, step, stop) where T
             last = steprange_last_empty(start, step, stop)
         else
             diff = stop - start
-            if T<:Signed && (diff > zero(diff)) != (stop > start)
```
Comment (mbauman) [inline base/range.jl]: Yes, I agree, this seems a little strange.  I understand why you need to remove this — and I agree it's a good thing to do, but I think we want to preserve the effective behavior here.

Perhaps we should detected (un)signed-ness based the behavior of `stop - start > zero(diff) && start - stop > zero(diff)`?

**Comment by tkoolen** [comment]
Alright, I've cleaned up `steprange_last`; should be a lot easier to see what's going on now.

**Comment by tkoolen** [comment]
Argh, that breaks `Date` ranges; I'll fix it.

**Comment by tkoolen** [comment]
Alright, to avoid making too many changes in this PR I opted to just use the old implementation of `steprange_last` in `Dates` (which should work fine for signed types like `TimeType`s) and to remove the `steprem` abstraction, as I didn't find it useful after the changes to `steprange_last`.

**Review by mbauman — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
Seems like it just needs a final review and decision by someone. Maybe @mbauman? What's the deal with the CI? Real failures or false alarms?

**Comment by mbauman** [comment]
I think this looks good — but it's touching such a sensitive piece of code I was hoping to get a second set of eyes on it.

Travis failure is a network thing in apt-get'ing dependencies. Win64 took 3 hours — so it looks like a timeout — but it's also reporting:
```
ERROR: LoadError: UndefVarError: stdin_monitor not defined
Stacktrace:
 [1] (::getfield(Main, Symbol("##44#50")))() at C:\projects\julia\julia-381ef4b43b\share\julia\test\runtests.jl:199
 [2] cd(::getfield(Main, Symbol("##44#50")), ::String) at .\file.jl:85
 [3] top-level scope
 [4] include at .\boot.jl:314 [inlined]
 [5] include_relative(::Module, ::String) at .\loading.jl:1070
 [6] include(::Module, ::String) at .\sysimg.jl:29
 [7] exec_options(::Base.JLOptions) at .\client.jl:264
 [8] _start() at .\client.jl:424
in expression starting at C:\projects\julia\julia-381ef4b43b\share\julia\test\runtests.jl:59
```

**Comment by tkoolen** [comment]
If there's any doubt about whether the CI failures are legitimate (I personally don't think the appveyor error is relevant, but still), could CI be rerun on this PR?

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by tkoolen** [comment]
If you prefer, I could add a fast path for `absstep == 1` in this PR, which would probably help with https://github.com/JuliaLang/julia/issues/26770.

**Comment by mbauman** [comment]
Let's see what Nanosoldier reports first.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27302#issuecomment-396681302) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/06fc70e696982acece137cd9291440d057688fcd/8f80bd2_vs_521dec9/report.md). cc @ararslan

**Comment by KristofferC** [comment]
Results are polluted by https://github.com/JuliaLang/julia/pull/27495#issuecomment-396806558

**Comment by mbauman** [comment]
@nanosoldier `runbenchmarks("array" || "collection" || "dates" || "find" || "random", vs = ":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/27302#issuecomment-396997989):
```
NanosoldierError: failed to run benchmarks against comparison commit: failed process: Process(`make -j3`, ProcessExited(2)) [2]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/0e3388cd395c5c07211db34af4b6a71b35269827/e46825a_vs_0658a11/report.md)
cc @ararslan

**Comment by tkoolen** [comment]
Please let me know if there's anything you need me to do.

**Comment by KristofferC** [comment]
Let's try again @nanosoldier `runbenchmarks("array" || "collection" || "dates" || "find" || "random", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27302#issuecomment-399050172) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/106c994476506e3305025bf25361958d0ca44949/1543c5f_vs_f90d674/report.md). cc @ararslan

**Comment by mbauman** [comment]
Great. CI is stale, just closing and re-opening to re-run it and then we can merge.  Thanks for your persistence here, @tkoolen!

**Comment by tkoolen** [comment]
Thank you! I know you guys are very busy right now, so I really appreciate you spending time on this PR.

**Comment by ararslan** [comment]
[Intervals](https://github.com/invenia/Intervals.jl) is using the `steprem` function to facilitate `StepRange{<:AbstractInterval}`. What's the recommended way to do it now, if defining `rem` doesn't make sense? DIY, as is done here with Dates?

**Comment by tkoolen** [comment]
I'm guessing it's this usage:

https://github.com/invenia/Intervals.jl/blob/3addff517786a1a832bb4e7b15ba65945d878c67/src/anchoredinterval.jl#L224-L226

Since the `AnchoredInterval` overload of `steprem` is just calling the `steprem` methods in `Base` with the `anchor`s as the arguments and [`Base.:-` is also defined just in terms of the anchors](https://github.com/invenia/Intervals.jl/blob/3addff517786a1a832bb4e7b15ba65945d878c67/src/anchoredinterval.jl#L202), can't you now just have a `steprange_last` method in Intervals that calls `steprange_last` on the anchors and converts to an `AnchoredInterval` at the end?

**Pull request merged**